### ISSUE TITLE
tsc --build --watch emitting unchanged files to disk

### ISSUE THREAD

**Issue by frags51** [microsoft/TypeScript#47309]
# Bug Report

Similar to  https://github.com/microsoft/TypeScript/issues/29927

### 🔎 Search Terms

 tsc --watch writing to disk too often

### 🕗 Version & Regression Information

Occurring in TS 3.9.5 (and also tried on TS 4.5, although we did get TS2571 errors in our code while using TS4.5.
We are using TS in a React-Native project. This issue is causing longer times to hot reload in RN.

tsconfig:
```
{
  "compilerOptions": {
    "module": "commonjs",
    "target": "es6",
    "jsx": "react",
    "moduleResolution": "node",
    "importHelpers": true,
    "experimentalDecorators": true,
    "resolveJsonModule": true,
    "sourceMap": false,
    "strictNullChecks": true,
    "noUnusedLocals": false,
    "noUnusedParameters": false,
    "forceConsistentCasingInFileNames": true,
    "noImplicitAny": true,
    "preserveConstEnums": true,
    "noErrorTruncation": true,
    "strict": true,
    "declaration": true,
    "allowJs": true,
    "skipLibCheck": true,
    "esModuleInterop": true
  },
  "compileOnSave": true
}
```

### 🙁 Actual behavior

On changing a single TS file, multiple JS files are written to disk. Can confirm that they need to be actually changed.

### 🙂 Expected behavior

Only the single (or files that actually need to be changed) should be written to disk.

**Comment by RyanCavanaugh** [comment]
This is not necessarily unexpected behavior for `--build` mode since it is a timestamp-based incremental build system. The content and structure of the files matters here and we'd need a complete repro to determine if there's a bug or not.

**Comment by DanielRosenwasser** [comment]
I've proposed a possible mitigation in https://github.com/microsoft/TypeScript/issues/46677

**Comment by frags51** [comment]
Project structure is as follows:
We have used Lerna to create a monorepo for our React Native app.

packages
&nbsp;|  - A (Root package, depends on B, C. `react-native start` is run from this package)
&nbsp;&nbsp;      | src/*.ts (Files from B, C are imported as `import Something from 'B/build/Something'` and not `'B/Something'`
&nbsp;&nbsp;       | app.js, index.js, tsconfig (at root level)
&nbsp;&nbsp;       | resources/images
 &nbsp;&nbsp;       | build/*.js (Emitted by tsc)
&nbsp;| - B
 &nbsp;&nbsp;      | src/*.ts
&nbsp;&nbsp;       | tsconfig (at root level)
&nbsp;&nbsp;       | build/*.js  (Emitted by tsc)
&nbsp;| - C
&nbsp;&nbsp;       | src/*.ts
&nbsp;&nbsp;       | tsconfig (at root level)
&nbsp;&nbsp;       | build/*.js  (Emitted by tsc)

tsc `--build` is needed so that changes in B, C also lead to a hot reload. If we remove `--build` still we see multiple JS files being emitted.

**Comment by DanielRosenwasser** [comment]
Sorry that you had to find mitigations for this. I believe that this is fixed now (since #46677 is fixed). If you still run into this in other projects, you can now try a nightly version of TypeScript and see if things are improved at all.
